Issue 14147 - Compiler crash on identical functions in a single module

9rnsr opened the pull request:
https://issues.dlang.org/show_bug.cgi?id=14147

Partial fix for [issue 2789](https://issues.dlang.org/show_bug.cgi?id=2789).

9rnsr commented:
Ready to merge.

WalterBright commented on src/func.c:
```diff
@@ -1222,6 +1223,49 @@ void FuncDeclaration::semantic(Scope *sc)
+            else if (AliasDeclaration *ad = s->isAliasDeclaration())
+            {
+                // todo
+                s = ad->overnext;
```
I suspect much of this function is duplicating what is in overloadApply(), but not handling all the cases.

9rnsr replied:
Now `overloadApply` is used to iterate function overloads.

WalterBright commented on src/mangle.c:
```diff
@@ -901,3 +901,26 @@ void mangleToBuffer(Expression *e, OutBuffer *buf)
     e->accept(&v);
 }
+
+void mangleToFuncSignature(OutBuffer *buf, FuncDeclaration *fd)
```
I don't understand why there is a same-only-different type signature routine. Why should this produce different results than tf->deco? And if it does produce something different, isn't something terribly wrong somewhere?

9rnsr replied:
As I commented to, overload conflict check does not need function return type. Even if two auto return functions defined in non-root module, their conflicts can be detected without their semantic3.

```d
auto foo() { ... }
auto foo() { ... }   // compiler can detect conflict, without semantic3 running
```

WalterBright commented on src/func.c:
```diff
@@ -1222,6 +1223,49 @@ void FuncDeclaration::semantic(Scope *sc)
+                    continue;
+
+                buf2.reset();
+                mangleToFuncSignature(&buf2, fd);
```
By computing the signature over and over, this can result in terrible compiler performance.

The  type `deco` system, on the other hand, produces a signature only once and uses a hash for lookup.

9rnsr replied:
As I already mentioned, `FuncDeclaratio.type.deco` is over-spec for conflict check. If we worry about performance, the `mangleToFuncSignature` result can be cached into `FuncDeclaration`.

yebblies commented:
I'm not sure the mangling approach is correct.  For example, this should be an error:
```
extern(D) void func() {}
extern(C) void func() {}
```

Is there some reason the check for matching (argument) signatures can't be done on `overloadInsert`?  Why is `semantic2` an appropriate place for this?

9rnsr commented:
`overloadInsert` is usually called from `addMember`. In that time, we cannot invoke semantic analysis to see correct function signatures because forward reference resolution is not yet ready (sibling scope members' `setScope` are not yet called).

In `semantic2`, function parameter types should be analyzed already in `semantic`. So it's the best place to see conflicts.

yebblies commented:
Is that for global functions too or just member functions?

Could it be done at the end of the parent symbol's semantic1 routine?

9rnsr commented:
> Is that for global functions too or just member functions?

For all functions in overloadable scopes. So includIng global functions.

> Could it be done at the end of the parent symbol's semantic1 routine?

Possible, but it needs modifying many places - Module.semantic, StructDeclaration.semantic, TemplateInstance.semantic, ... Instead do it in FuncDeclatation.semantic2 is much more sImple.

9rnsr commented:
Rebased onto master, and add support to detect `extern (C)` function conflicts.

yebblies commented:
Still doesn't error on this?
```
extern(D) void func() {}
extern(C) void func() {}
```

9rnsr commented:
> Still doesn't error on this?

No, now it will raise conflict error. See the case `f_MixExtern1` in `fail2789.d`.

yebblies commented on src/dmangle.d:
```diff
@@ -862,3 +862,32 @@ extern (C++) void mangleToBuffer(Expression e, OutBuffer* buf)
+
+    //if (fd.needThis() || fd.isNested())
+    //    buf.writeByte(Type.needThisPrefix());
+    MODtoDecoBuffer(buf, tf.mod);
```
Why is this necessary?  Which test cases are affected by this?

9rnsr replied:
`void foo()` and `void foo() const` can be overloaded each other.

yebblies replied:
Huh, for some reason I thought this was only for non-member functions, and member functions were checked elsewhere.

yebblies commented on src/func.d:
```diff
@@ -1286,12 +1287,99 @@ public:
+
+            // Always starts the lookup from 'this', because the conflicts with
+            // previous overloads are already reported.
+            auto f1 = this;
```
Why have a copy in `f1`?  Wouldn't using `this` make the code easier to follow?

9rnsr replied:
It's just for the code reading. To me `f1` vs `f2` is easier to understand than `this` vs `f2`.

yebblies commented on src/func.d:
```diff
@@ -1286,12 +1287,99 @@ public:
+                if (strcmp(s1, s2) == 0)
+                {
+                    auto tf2 = cast(TypeFunction)f2.type;
+                    f2.error("%s conflicts with at %s",
```
This need a grammatical fix - maybe `"%s conflicts with %s at %s"` or `"%s conflicts with another declaration at %s"` if you want to avoid naming the second func.

yebblies replied:
`"overloaded functions have the same argument lists"` and then list overloads?

9rnsr replied:
I'll update the message to `"%s conflicts with previous declaration at %s"`.

yebblies replied:
Even better.

yebblies commented on src/func.d:
```diff
@@ -1286,12 +1287,99 @@ public:
+                    f1.toParent2() == f2.toParent2() &&
+                    f1.linkage == LINKc && f2.linkage == LINKc)
+                {
+                    /* Allow the hack that is actually used in druntime,
```
Could you please file a bug about this issue so we don't forget about it?

9rnsr replied:
OK, will do.

9rnsr replied:
Done. https://issues.dlang.org/show_bug.cgi?id=15217

yebblies commented on src/func.d:
```diff
@@ -1286,12 +1287,99 @@ public:
+
+                    auto tf1 = cast(TypeFunction)f1.type;
+                    auto tf2 = cast(TypeFunction)f2.type;
+                    f2.error("%s conflicts with at %s because of extern (C)",
```
`"extern (C) %s conflicts with another function at %s"`?
Or possibly
`"extern (C) functions cannot be overloaded"` and then list the conflicting overloads (with locations).  I think this would be best for avoiding confusion.

9rnsr replied:
I'll update the message to: `"%s cannot be overloaded with another extern (C) function at %s"`.

yebblies replied:
Great.  And please include `extern (Windows)` etc

9rnsr replied:
`extern (Windows)` needs ... more work. If two `exntern (Windows) void foo(int*)` and `void foo(void*)` exist, they don't conflicts on semantic overload level, but they will have same link name `_foo@4` *only in 32bit codegen*.
On the other hand, in win64 `extern (Windows)` is equivalent with `extern (C)`.

Please give me some time...

yebblies replied:
>`extern (Windows)` needs ... more work. If two `exntern (Windows) void foo(int*)` and `void foo(void*)` exist, they don't conflicts on semantic overload level, but they will have same link name `_foo@4` *only in 32bit codegen*.

I don't think it matters that they might have a different linker symbol name.  They're not designed for overloading, and I think it would be best to simply disallow it.

> Please give me some time...

Of course!

9rnsr replied:
> They're not designed for overloading, and I think it would be best to simply disallow it.

OK, make sense.

yebblies commented:
I think all linkages except C++ and D should not be able to be overloaded.

yebblies commented on src/func.d:
```diff
@@ -1286,12 +1287,99 @@ public:
+            overloadApply(f1, (Dsymbol s)
+            {
+                auto f2 = s.isFuncDeclaration();
+                if (!f2|| f1 == f2 || f2.errors)
```
Spacing around `||`

9rnsr replied:
Will fix.

yebblies commented on src/dmangle.d:
```diff
@@ -862,3 +862,32 @@ extern (C++) void mangleToBuffer(Expression e, OutBuffer* buf)
+
+    scope Mangler v = new Mangler(buf);
+
+    //if (fd.needThis() || fd.isNested())
```
Comments can be deleted now?

9rnsr commented:
> I think all linkages except C++ and D should not be able to be overloaded.

I think one `extern(C)` and one or more `extern (D)` functions can be overloaded.

yebblies commented:
> I think one `extern(C)` and one or more `extern (D)` functions can be overloaded.

Yes you're right.  I'm sure I've done that before.

9rnsr committed:
fix Issue 2789 - Functions overloads are not checked for conflicts

9rnsr committed:
Treat LINKwindows, LINKpascal, and LINKobjc as equivalent with LINKc

They conflict each other.

9rnsr committed:
Allow an overload declaration hack, which is actually used in druntime

If two extern (C) functions are just declared with different signatures,
they don't conflict.

 extern(C):
 alias sigfn_t  = void function(int);
 alias sigfn_t2 = void function(int) nothrow @nogc;
 sigfn_t  bsd_signal(int sig, sigfn_t  func);
 sigfn_t2 bsd_signal(int sig, sigfn_t2 func) nothrow @nogc;  // no error

This behavior must be reconsidered in the future.

9rnsr committed:
fix Issue 14147 - Compiler crash on identical functions in a single module

9rnsr commented:
Now `extern (Windows)` and `extern (C)` conflict each other.

CyberShadow commented:
>  Allow an overload declaration hack, which is actually used in druntime 

That use was added very recently, maybe we should fix Druntime instead?

yebblies commented:
What about `extern (Pascal)`?

9rnsr commented:
I don't know well about `extern (Pascal)`.

9rnsr commented:
@CyberShadow I think, no.

yebblies commented:
`extern (Pascal)` is similar to `extern (Windows)`, and does not support overloading.

9rnsr commented:
> does not support overloading

Where is it documented? I couldn't find any specification about `extern (Pascal)`, so I don't want to add any special behaviors for that.

yebblies commented:
Pascal is an older variant of stdcall.

9rnsr commented:
OK, now `LINKc`, `LINKwindows`, `LINKpascal`, and `LINKobjc` conflict each other.

yebblies commented:
> @CyberShadow I think, no.

Why shouldn't druntime be fixed?

9rnsr commented:
[Issue 15217](https://issues.dlang.org/show_bug.cgi?id=15217) looks intentional to propagate delegate argument attributes to the function call. Currently there's not alternative way to do it, so I think we should leave it.

wilzbach commented:
Rebased to #7577
